src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetMetadataStep.java

```diff
@@ -36,9 +37,16 @@ public CreateDatasetMetadataStep(DatasetDao datasetDao, DatasetRequestModel data
     public StepResult doStep(FlightContext context) {
         try {
             Dataset newDataset = DatasetUtils.convertRequestWithGeneratedNames(datasetRequest);
-            UUID datasetId = datasetDao.createAndLock(newDataset, context.getFlightId());
+
             FlightMap workingMap = context.getWorkingMap();
+            UUID datasetId = UUID.randomUUID();
+            Instant createdDate = Instant.now();
             workingMap.put(DatasetWorkingMapKeys.DATASET_ID, datasetId);
+            newDataset
```

Comment:
This should be split into two steps. A new step creates the id and puts it in the working map. That's it. Then it will be persisted in Stairway.

The metadata step then reads that id and tries the createAndLock. The undo does what you have written, but since the id step is guaranteed to complete, it will know for sure that there is an id in the working map.